File: Neetcode/KokoBanana.py
from math import ceil
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        if len(piles) > h:
            return -1
        if len(piles) == h:
            return max(piles)
        else:
            left = 1
            right = max(piles)
            ans = right + 1
            while left <= right:
                mid = left + (right - left)//2
                time = 0
                for p in piles:
                    time += ceil(p/mid)
                if time > h:
                    left = mid + 1
                else:
                    ans = min(ans, mid)
                    right = mid - 1
            return ans

File: Neetcode/KokoBanana.cpp
#include <iomanip>
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        std::cout << std::fixed << std::setprecision(10);
        int left = 1;
        int right = -1;
        for (int i=0; i<piles.size(); i++){
            right = max(right, piles[i]);
        }
        int mid = -1;
        int minspeed = right+1;
        while (left <= right){
            mid = left + (right - left)/2;
            long long int hours = 0;
            // cout << mid << "\n";
            for (int i=0; i<piles.size(); i++){
                // cout << static_cast<long double>(piles[i])/static_cast<long double>(mid) << "\n";
                hours += ceil(static_cast<long double>(piles[i])/static_cast<long double>(mid));
                // cout << hours << "\n";
            }
            if (hours <= h){
                minspeed = min(minspeed, mid);
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }
        return minspeed;
    }
};
